### PR TITLE
Flush initial styled-jsx in gIP first in concurrent rendering

### PR DESCRIPTION
**Pull request by huozhi**
### Bug

If we're flushing styled-jsx collected styles from both Document.gIP and flush effects in react 18, we will get duplicated styled-jsx in both head (from gIP) and body (from flush effects). But when there's jsx styles under `Suspense`, gIP cannot collect all the ones (suspended styles are missing).

### Solution

flush initial styles in gIP, keep other suspended styles in body

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 56375e9ea3cbf11c2de014bd2b75a490f3283f06

`yarn testheadless test/e2e/styled-jsx/index.test.ts` 
- styled-jsx > should contain styled-jsx styles in html
- styled-jsx pnpm > should contain styled-jsx styles in html

<details>
<summary>Expand output</summary>

  ● styled-jsx › should contain styled-jsx styles in html

    expect(received).toContain(expected) // indexOf

    Expected substring: "color:red"
    Received string:    "body{display:none}body{display:block}"

      34 |       const html = await renderViaHTTP(next.url, '/')
      35 |       const $ = cheerio.load(html)
    > 36 |       expect($('head').text()).toContain('color:red')
         |                                ^
      37 |     })
      38 |   })
      39 | }

      at Object.<anonymous> (e2e/styled-jsx/index.test.ts:36:32)

  ● styled-jsx pnpm › should contain styled-jsx styles in html

    expect(received).toContain(expected) // indexOf

    Expected substring: "color:red"
    Received string:    "body{display:none}body{display:block}"

      34 |       const html = await renderViaHTTP(next.url, '/')
      35 |       const $ = cheerio.load(html)
    > 36 |       expect($('head').text()).toContain('color:red')
         |                                ^
      37 |     })
      38 |   })
      39 | }

      at Object.<anonymous> (e2e/styled-jsx/index.test.ts:36:32)
          at runMicrotasks (<anonymous>)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | huozhi/next.js fix/duplicated-styled-jsx-styles | Change |
| - | - | - | - |
| buildDuration | 17.9s | 17.6s | -294ms |
| buildDurationCached | 7s | 6.9s | -105ms |
| nodeModulesSize | 463 MB | 463 MB | ⚠️ +1.14 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | huozhi/next.js fix/duplicated-styled-jsx-styles | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 4.26 | 4.178 | -0.08 |
| / avg req/sec | 586.84 | 598.39 | +11.55 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.58 | 1.594 | ⚠️ +0.01 |
| /error-in-render avg req/sec | 1582.06 | 1568.19 | ⚠️ -13.87 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js fix/duplicated-styled-jsx-styles | Change |
| - | - | - | - |
| 925.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 28.6 kB | 28.6 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 72.3 kB | 72.3 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js fix/duplicated-styled-jsx-styles | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js fix/duplicated-styled-jsx-styles | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 193 B | 193 B | ✓ |
| amp-HASH.js gzip | 308 B | 308 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 3.08 kB | 3.08 kB | ✓ |
| head-HASH.js gzip | 359 B | 359 B | ✓ |
| hooks-HASH.js gzip | 920 B | 920 B | ✓ |
| image-HASH.js gzip | 5.71 kB | 5.71 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.64 kB | 2.64 kB | ✓ |
| routerDirect..HASH.js gzip | 320 B | 320 B | ✓ |
| script-HASH.js gzip | 391 B | 391 B | ✓ |
| withRouter-HASH.js gzip | 318 B | 318 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16.3 kB | 16.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js fix/duplicated-styled-jsx-styles | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js fix/duplicated-styled-jsx-styles | Change |
| - | - | - | - |
| index.html gzip | 533 B | 533 B | ✓ |
| link.html gzip | 547 B | 547 B | ✓ |
| withRouter.html gzip | 529 B | 529 B | ✓ |
| Overall change | 1.61 kB | 1.61 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | huozhi/next.js fix/duplicated-styled-jsx-styles | Change |
| - | - | - | - |
| buildDuration | 20.4s | 20.4s | -1ms |
| buildDurationCached | 7s | 7.3s | ⚠️ +301ms |
| nodeModulesSize | 463 MB | 463 MB | ⚠️ +1.14 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | huozhi/next.js fix/duplicated-styled-jsx-styles | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 4.295 | 4.378 | ⚠️ +0.08 |
| / avg req/sec | 582.03 | 571.05 | ⚠️ -10.98 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.572 | 1.658 | ⚠️ +0.09 |
| /error-in-render avg req/sec | 1590.78 | 1507.71 | ⚠️ -83.07 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js fix/duplicated-styled-jsx-styles | Change |
| - | - | - | - |
| 925.HASH.js gzip | 178 B | 178 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 29 kB | 29 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 72.9 kB | 72.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js fix/duplicated-styled-jsx-styles | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js fix/duplicated-styled-jsx-styles | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 179 B | 179 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 324 B | 324 B | ✓ |
| dynamic-HASH.js gzip | 3.08 kB | 3.08 kB | ✓ |
| head-HASH.js gzip | 357 B | 357 B | ✓ |
| hooks-HASH.js gzip | 921 B | 921 B | ✓ |
| image-HASH.js gzip | 5.76 kB | 5.76 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 2.76 kB | 2.76 kB | ✓ |
| routerDirect..HASH.js gzip | 322 B | 322 B | ✓ |
| script-HASH.js gzip | 392 B | 392 B | ✓ |
| withRouter-HASH.js gzip | 317 B | 317 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16.5 kB | 16.5 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js fix/duplicated-styled-jsx-styles | Change |
| - | - | - | - |
| _buildManifest.js gzip | 458 B | 458 B | ✓ |
| Overall change | 458 B | 458 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js fix/duplicated-styled-jsx-styles | Change |
| - | - | - | - |
| index.html gzip | 528 B | 528 B | ✓ |
| link.html gzip | 541 B | 541 B | ✓ |
| withRouter.html gzip | 525 B | 525 B | ✓ |
| Overall change | 1.59 kB | 1.59 kB | ✓ |

</details>

</details>
Commit: 56375e9ea3cbf11c2de014bd2b75a490f3283f06

File: packages/next/server/render.tsx [hunk 1]
```diff
@@ -1539,18 +1551,11 @@ export async function renderToHTML(
           renderStream,
           suffix,
           dataStream: serverComponentsInlinedTransformStream?.readable,
-          generateStaticHTML:
-            generateStaticHTML || !process.env.__NEXT_REACT_ROOT,
+          generateStaticHTML,
```
Comment (shuding) [inline packages/next/server/render.tsx:1554]: Is it true that this will always be enabled for React 17 and below?

Comment (huozhi) [reply packages/next/server/render.tsx:1547]: There's a condition branch checking `__NEXT_REACT_ROOT` above, if it's using react 17, will go to another `if` branch. Here react root is always `true` then `|| !process.env.__NEXT_REACT_ROOT` is always `false`.

I also added some comments for `__NEXT_REACT_ROOT ` condition in the if-else branch

**Review by shuding — DISMISSED** [review]

**Review by shuding — APPROVED** [review]

**Pull request merged**